BR: refine BR docs (#8628)
pingcap/docs: br/backup-and-restore-overview.md

- **Restore a database or a table from backup**: You can restore part of a specific database or table. During the process, BR will filter out unnecessary data. For details, see [Restore a database or a table](/br/br-usage-restore.md#restore-a-database-or-a-table).

#### Impact on performance

Data restoration is performed at a scalable speed. Generally, the speed is 100 MB/s per TiKV node. BR only supports restoring data to a new cluster and uses the resources of the target cluster as much as possible. For more details, see [Restoration performance and impact](/br/br-usage-restore.md#restoration-performance-and-impact).

## Before you use BR

Before you use BR, pay attention to its usage restrictions, compatibility, and other considerations.

### Usage restrictions

This section describes usage restrictions of BR.

#### Unsupported scenarios

When BR restores data to the upstream cluster of TiCDC or TiDB Binlog, TiCDC or TiDB Binlog cannot replicate the restored data to the downstream cluster.

#### Compatibility

The compatibility issues of BR and a TiDB cluster are as follows:

- There is a cross-version compatibility issue:

>     BR (< v5.4.0) cannot restore `charset=GBK` tables. In the meantime, no version of BR supports restoring `charset=GBK` tables to a TiDB cluster earlier than v5.4.0.

Reviewer: ```suggestion
    Before v5.4.0, BR cannot restore tables with `charset=GBK`. At the same time, no version of BR supports restoring `charset=GBK` tables to a TiDB cluster earlier than v5.4.0.
```

---
title: BR Overview
summary: Learn about the definition and functions of BR.
aliases: ['/docs/dev/br/backup-and-restore-tool/','/docs/dev/reference/tools/br/br/','/docs/dev/how-to/maintain/backup-and-restore/br/','/tidb/dev/backup-and-restore-tool']
---

# BR Overview

[BR](https://github.com/pingcap/br) (Backup & Restore) is a command-line tool for **distributed backup and restoration** of the TiDB cluster data. In addition to regular backup and restoration, you can also use BR for large-scale data migration as long as compatibility is ensured.

This document describes BR's architecture, functions, and usage tips.

## BR architecture

BR sends a backup or restoration command to each TiKV node. After receiving the command, TiKV performs the corresponding backup or restoration operation.

> Each TiKV node has a path in which the backup files generated in the backup operation are stored and from which the stored backup files are read during the restoration.

Reviewer 2: ```suggestion
Each TiKV node needs to access the storage path during backup and restoration.
```

![br-arch](/media/br-arch.png)

For detailed information about the BR design, see [BR Design Principles](/br/backup-and-restore-design.md).

## BR features

This section describes BR features and performance indicators.

### Back up TiDB cluster data

This section describes backup features, impact of backup, and backup file types.

#### Backup features

> - **Back up cluster snapshots**: A snapshot of a TiDB cluster contains transactionally consistent data at a specific time. You can back up snapshot data of a TiDB cluster using BR. For details, see [Back up TiDB cluster snapshots](/br/br-usage-backup.md#back-up-tidb-cluster-snapshots).

Reviewer: this one is missing

> - **Back up cluster snapshots**: A snapshot of a TiDB cluster contains transactionally consistent data at a specific time. You can back up snapshot data of a TiDB cluster using BR. For details, see [Back up TiDB cluster snapshots](/br/br-usage-backup.md#back-up-tidb-cluster-snapshots).

Author: Changed the Chinese in https://github.com/pingcap/docs-cn/pull/10099/files#diff-5f5519a7f44e95526e3253e724fdc87ff3e57461ceae2bfd6317dc96775fe3fcR27

> - **Back up cluster snapshots**: A snapshot of a TiDB cluster contains only the latest and transactionally consistent data at a specific time. You can back up snapshot data of a TiDB cluster using BR. For details, see [Back up TiDB cluster snapshots](/br/br-usage-backup.md#back-up-tidb-cluster-snapshots).

Reviewer 2: ```suggestion
- **Back up cluster snapshots**: A snapshot of a TiDB cluster contains transactionally consistent data at a specific time. You can back up snapshot data of a TiDB cluster using BR. For details, see [Back up TiDB cluster snapshots](/br/br-usage-backup.md#back-up-tidb-cluster-snapshots).
```

> - **Back up incremental data**: Incremental data of a TiDB cluster is differentiated data between the latest snapshot and the previous snapshot. Incremental data is smaller compared with full data and it is a supplementary to snapshot backup, which reduces the volume of backup data. For details, see [Back up incremental data](/br/br-usage-backup.md#back-up-incremental-data).

Reviewer: ```suggestion
- **Back up incremental data**: The incremental data of a TiDB cluster represents changes between the latest snapshot and the previous snapshot. Incremental data is smaller in size compared with full data, and can be used together with snapshot backup, which reduces the volume of backup data. For details, see [Back up incremental data](/br/br-usage-backup.md#back-up-incremental-data).
```

> - **Back up a database or a table**: In addition to snapshot and incremental data backup, BR supports backing up a database or table and filtering out unnecessary data. For details, see [Back up a database or table](/br/br-usage-backup.md#back-up-a-database-or-a-table).

Reviewer: ```suggestion
- **Back up a specific database or table**: On top of snapshot and incremental data backup, BR supports backing up a specific database or table and filtering out unnecessary data. For details, see [Back up a database or table](/br/br-usage-backup.md#back-up-a-database-or-a-table).
```

> - **Encrypt backup data**: BR supports backup data encryption and amazon S3 server-side encryption. You can select an encryption method as required. For details, see [Encrypt backup data](/br/br-usage-backup.md#encrypt-backup-data).

Reviewer: ```suggestion
- **Encrypt backup data**: BR supports data encryption in the backup end or in the Amazon S3 storage server end. You can select an encryption method as needed. For details, see [Encrypt backup data](/br/br-usage-backup.md#encrypt-backup-data).
```

Reviewer: missing

Author: Changed the Chinese in https://github.com/pingcap/docs-cn/pull/10099/files#diff-5f5519a7f44e95526e3253e724fdc87ff3e57461ceae2bfd6317dc96775fe3fcR30

Reviewer: Comment updated

Author: server-side encryption seems like a fixed term. https://docs.aws.amazon.com/AmazonS3/latest/userguide/serv-side-encryption.html

> - **Encrypt backup data**: BR supports encrypting data at the backup source or the storage end when data is backed up to Amazon S3. You can select an encryption method as required. For details, see [Encrypt backup data](/br/br-usage-backup.md#encrypt-backup-data).

Reviewer 2: ```suggestion
- **Encrypt backup data**: BR supports backup data encryption and amazon S3 server-side encryption. You can select an encryption method as required. For details, see [Encrypt backup data](/br/br-usage-backup.md#encrypt-backup-data).
```

Author: agree with this modification. Maybe we should synchronize the Chinese as well.

#### Impact on performance

> The impact of backup on a TiDB cluster is kept below 20%, and this value can be reduced to 10% or less with reasonable configuration of the TiDB cluster. The backup speed of a TiKV node is scalable and ranges from 50 MB/s to 100 MB/s. For more information, see [Backup performance and impact](/br/br-usage-backup.md#backup-performance-and-impact).

Reviewer: ```suggestion
The impact of backup on a TiDB cluster is kept below 20%, and this value can be reduced to 10% or less with the proper configuration of the TiDB cluster. The backup speed of a TiKV node is scalable and ranges from 50 MB/s to 100 MB/s. For more information, see [Backup performance and impact](/br/br-usage-backup.md#backup-performance-and-impact).
```

> #### Storage of backup data

Reviewer: ```suggestion
#### Storage types of backup data
```

> BR supports backing up data to Amazon S3, Google Cloud Storage, Azure Blob Storage, NFS, and other S3-compatible file storage services. For details, see [Back up data to external storage](/br/br-usage-backup.md#back-up-data-to-external-storage).

Reviewer: ```suggestion
BR supports backing up data to Amazon S3, Google Cloud Storage, Azure Blob Storage, NFS, and other S3-compatible file storage services. For details, see [Back up data to external storages](/br/br-usage-backup.md#back-up-data-to-external-storage).
```

### Restore TiDB cluster data

This section describes restoration features and impact on performance.

#### Restoration features

> - **Restore cluster snapshot backup**: You can restore snapshot backup data to a new cluster. For details, see [Restore TiDB cluster snapshots](/br/br-usage-restore.md#restore-tidb-cluster-snapshots).

Reviewer: ```suggestion
- **Restore snapshot backup**: You can restore snapshot backup data to a new cluster. For details, see [Restore TiDB cluster snapshots](/br/br-usage-restore.md#restore-tidb-cluster-snapshots).
```

> - **Restore cluster snapshot backup**: You can restore cluster data to that of a snapshot backup in an empty cluster at a certain time. For details, see [Restore TiDB cluster snapshots](/br/br-usage-restore.md#restore-tidb-cluster-snapshots).

Reviewer 2: ```suggestion
- **Restore cluster snapshot backup**: You can restore backup data to an new cluster. For details, see [Restore TiDB cluster snapshots](/br/br-usage-restore.md#restore-tidb-cluster-snapshots).
```

> - **Restore cluster incremental backup**: You can restore the incremental backup data to a cluster. For details, see [Restore incremental backup](/br/br-usage-restore.md#restore-incremental-data).

Reviewer: ```suggestion
- **Restore incremental backup**: You can restore the incremental backup data to a cluster. For details, see [Restore incremental backup](/br/br-usage-restore.md#restore-incremental-data).
```

> - **Restore cluster incremental backup**: You can restore cluster data to that of an incremental backup at a certain time. For details, see [Restore incremental backup](/br/br-usage-restore.md#restore-incremental-data).

Reviewer 2: ```suggestion
- **Restore cluster incremental backup**: You can restore the incremental backup data to a cluster. For details, see [Restore incremental backup](/br/br-usage-restore.md#restore-incremental-data).
```

> - **Restore a database or a table from backup**: You can restore part of a specific database or table. During the process, you need to filter out unnecessary data. For details, see [Restore a database or a table](/br/br-usage-restore.md#restore-a-database-or-a-table).

Reviewer 2: ```suggestion
- **Restore a database or a table from backup**: You can restore part of a specific database or table. During the process, BR will filter out unnecessary data. For details, see [Restore a database or a table](/br/br-usage-restore.md#restore-a-database-or-a-table).
```

#### Impact on performance

> Restoration proceeds with a scalable speed. Generally, the speed is 100 MB/s per TiKV node. BR only supports restoring data to a new cluster and uses resources of the target cluster as much as possible. For more details, see [Restoration performance and impact](/br/br-usage-restore.md#restoration-performance-and-impact).

Reviewer: ```suggestion
Data restoration is performed at a scalable speed. Generally, the speed is 100 MB/s per TiKV node. BR only supports restoring data to a new cluster and uses the resources of the target cluster as much as possible. For more details, see [Restoration performance and impact](/br/br-usage-restore.md#restoration-performance-and-impact).
```

## Before you use BR

Before you use BR, pay attention to its usage restrictions, compatibility, and other considerations.

### Usage restrictions

This section describes usage restrictions of BR.

> #### Unsupported scenario

Reviewer: ```suggestion
#### Unsupported scenarios
```

When BR restores data to the upstream cluster of TiCDC or TiDB Binlog, TiCDC or TiDB Binlog cannot replicate the restored data to the downstream cluster.

> #### Compatibility

Reviewer: Why not reuse the deleted content in br/backup-and-restore-tools.md? The deleted content is not only fully applicable here but also looks better in many details.

The compatibility issues of BR and a TiDB cluster are as follows:

> - There are some cross-version compatibility issues:

Reviewer: Currently there is only one issue.

>     BR(< v5.4.0) cannot restore `charset=GBK` tables. In the meantime, no version of BR supports restoring `charset=GBK` tables to a TiDB cluster earlier than v5.4.0.

Reviewer: Please avoid using `<` because it might be parsed as a tag.
```suggestion
    BR (< v5.4.0) cannot restore `charset=GBK` tables. In the meantime, no version of BR supports restoring `charset=GBK` tables to a TiDB cluster earlier than v5.4.0.
```

> 

Reviewer: This part is missing from Chinese. Is that intended?
![image](https://user-images.githubusercontent.com/41534398/173780480-a052c9b8-b4e4-4adf-a688-1697197d1aaf.png)

Author: Add this part to the FAQ after confirmation with Ian: https://github.com/pingcap/docs-cn/pull/10099/files#diff-5d65e67605cdab9570939131909b470ad54050d43dc181aacc0a832850ccbacaR230

> - Some versions of BR are not compatible with the interface of the TiDB cluster.

Reviewer 2: ```suggestion
- There are some cross-version compatibility issues:
```

>     Before v5.4.0, BR cannot restore `charset=GBK` tables. In the meantime, no version of BR supports restoring `charset=GBK` tables to a TiDB cluster earlier than v5.4.0.

Reviewer 2: ```suggestion
    BR(< v5.4.0) cannot restore `charset=GBK` tables. In the meantime, no version of BR supports restoring `charset=GBK` tables to a TiDB cluster whose version is earlier than v5.4.0.
```

|  ----  |  ----  | ---- | ---- |
| Use BR nightly to back up TiDB nightly | ✅ | ✅ | ❌ (If a table with the primary key of the non-integer clustered index type is restored to a TiDB v4.0 cluster, BR will cause data error without warning.) |
| Use BR v5.0 to back up TiDB v5.0 | ✅ | ✅ | ❌ (If a table with the primary key of the non-integer clustered index type is restored to a TiDB v4.0 cluster, BR will cause data error without warning.) |
| Use BR v4.0 to back up TiDB v4.0 | ✅ | ✅  | ✅ (If TiKV >= v4.0.0-rc.1, and if BR contains the [#233](https://github.com/pingcap/br/pull/233) bug fix and TiKV does not contain the [#7241](https://github.com/tikv/tikv/pull/7241) bug fix, BR will cause the TiKV node to restart.)|
| Use BR nightly or v5.0 to back up TiDB v4.0 | ❌ (If the TiDB version is earlier than v4.0.9, the [#609](https://github.com/pingcap/br/issues/609) issue might occur.)| ❌ (If the TiDB version is earlier than v4.0.9, the [#609](https://github.com/pingcap/br/issues/609) issue might occur.)| ❌ (If the TiDB version is earlier than v4.0.9, the [#609](https://github.com/pingcap/br/issues/609) issue might occur.) |

> 

Reviewer: This whole part is missing from the Chinese version. Is that intended?
![image](https://user-images.githubusercontent.com/41534398/173780207-a45883e3-c98b-4ee3-8f30-7f549253e611.png)

Author: According to confirmation with Ian, this information is covered in this table item: https://github.com/pingcap/docs/pull/8628/files#diff-5f5519a7f44e95526e3253e724fdc87ff3e57461ceae2bfd6317dc96775fe3fcR79.
![image](https://user-images.githubusercontent.com/89768198/175848912-22d10477-473e-414c-847f-d6f95066c05c.png)

Author: Fixed in https://github.com/pingcap/docs/pull/8628/commits/0d1ff8980803e77e4c15c6b78321add372489338#diff-5d65e67605cdab9570939131909b470ad54050d43dc181aacc0a832850ccbacaR230.
